File: engine/src/flow/layers/texture_layer.cc
#include "flow/layers/texture_layer.h"

#include "flow/texture.h"

namespace uiwidgets {

TextureLayer::TextureLayer(const SkPoint& offset, const SkSize& size,
                           int64_t texture_id, bool freeze)
    : offset_(offset), size_(size), texture_id_(texture_id), freeze_(freeze) {}

void TextureLayer::Preroll(PrerollContext* context, const SkMatrix& matrix) {
  TRACE_EVENT0("uiwidgets", "TextureLayer::Preroll");

  set_paint_bounds(SkRect::MakeXYWH(offset_.x(), offset_.y(), size_.width(),
                                    size_.height()));
}

void TextureLayer::Paint(PaintContext& context) const {
  TRACE_EVENT0("uiwidgets", "TextureLayer::Paint");

  std::shared_ptr<Texture> texture =
      context.texture_registry.GetTexture(texture_id_);
  if (!texture) {
    TRACE_EVENT_INSTANT0("uiwidgets", "null texture");
    return;
  }
  texture->Paint(*context.leaf_nodes_canvas, paint_bounds(), freeze_,
                 context.gr_context);
}

}  // namespace uiwidgets

File: com.unity.uiwidgets/Runtime/ui/renderer/compositeCanvas/flow/texture_layer.cs
using Unity.UIWidgets.foundation;
using Unity.UIWidgets.ui;
using UnityEngine;
using Rect = Unity.UIWidgets.ui.Rect;

namespace Unity.UIWidgets.flow {
    public class TextureLayer : Layer {
        Offset _offset = Offset.zero;

        public Offset offset {
            set { _offset = value ?? Offset.zero; }
        }

        Size _size;

        public Size size {
            set { _size = value; }
        }

        Texture _texture;

        public Texture texture {
            set { _texture = value; }
        }

        bool _freeze = false;

        public bool freeze {
            set { _freeze = value; }
        }

        public override void preroll(PrerollContext context, Matrix3 matrix) {
            paintBounds = Rect.fromLTWH(
                _offset.dx, _offset.dy, _size.width, _size.height);
        }

        public override void paint(PaintContext context) {
            D.assert(needsPainting);

            if (_texture == null) {
                return;
            }

            var image = new Image(_texture, noDispose: true);

            var canvas = context.canvas;
            canvas.drawImageRect(image, paintBounds, new Paint());
        }
    }
}